### PR TITLE
fix: small security hardening (invite, DMARC, repo paths)

### PR DESCRIPTION
**Pull request by jonastemplestein**
## Summary

Three small, high-ROI security fixes found by a CodeCrucible scan of `apps/os` + `apps/auth`:

1. **`organization.createInvite`** — require owner (or platform admin) to invite as `role: "owner"`, matching the existing guards on `updateMemberRole` / `removeMember`. Non-owner org admins could previously mint an owner invite for themselves.
2. **`dmarcPasses`** — require `mx.cloudflare.net` on the same Authentication-Results line as `dmarc=pass`, so a self-forged `Authentication-Results: evil; dmarc=pass` no longer satisfies the allowlist gate.
3. **`normalizeRepoFilePath`** — reject bare `..` and any path segment that is `..` / `.` / empty (e.g. `foo/..`), which the previous prefix/substring checks missed.

Not in this PR (larger work): Slack `!` bang-command injection, webhook-post SSRF, worker build-cache IDOR.

## Test plan

- [x] `pnpm --dir apps/os exec vitest run src/domains/email/utils.test.ts` (includes new DMARC spoof rejection cases)
- [ ] Confirm createInvite as org admin with `role: "owner"` returns FORBIDDEN; as owner still works
- [ ] Confirm `commitFiles` / `edit` reject path `".."` and `"foo/.."`

---

> [!NOTE]
> **Medium Risk**
> Changes touch authorization (owner invites), inbound email acceptance (DMARC parsing), and repo mutation path validation; behavior is tightened with tests, but mis-parsed DMARC or overly strict paths could block legitimate mail or commits if edge cases were missed.
> 
> **Overview**
> Three targeted security fixes across auth and OS.
> 
> **Organization invites** — `createInvite` now returns **FORBIDDEN** when a non-owner org admin (or anyone without platform admin) tries to invite with `role: "owner"`, aligned with `updateMemberRole` and `removeMember`.
> 
> **Inbound email DMARC** — `dmarcPasses` no longer treats any `dmarc=pass` substring as sufficient. It splits `Authentication-Results` (newlines and careful comma joins from `Headers.get`), finds **Cloudflare** records (`mx.cloudflare.net` with optional version, rejecting prefix forgeries), and requires `dmarc=pass` on the **last** such record so forged headers and stale passes cannot bypass the allowlist gate.
> 
> **Repo file paths** — `normalizeRepoFilePath` now rejects bare `..`, empty segments, `.` / `..` in any segment (e.g. `foo/..`, `foo/./bar`), closing gaps left by earlier prefix-only checks on `commitFiles` / `edit` / `readFile`.
> 
> Tests add broad DMARC spoof and join scenarios in `utils.test.ts`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d98251cd89026a934f3a9f3987203d5e150578cd. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

| Group | Lines | Significant |
| --- | ---: | ---: |
| Product | +59 -13 | +31 -5 🟩🟩🟩🟩🟥 |
| Tests | +44 -1 | +32 -1 🟩🟩🟩🟩🟩 |
| **Total** | **+103 -14** | **+63 -6** 🟩🟩🟩🟩🟩 |

<sub>Lines counts every changed line; Significant ignores blank lines, JS comments, and TypeScript lines with no runtime output. Between b6e46ef and 4f700d2, bucketed first-match-wins into the groups defined in `scripts/ci/loc-report.ts`.</sub>

## Environment Config Lease

<details>
<summary>No preview slot recorded.</summary>

| app | status | commit | preview | size (gzip) | deploy duration | test duration | retries | cleanup duration | workflow run | updated | summary |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Auth | released | `d98251c` | [https://auth.iterate-preview-4.com](https://auth.iterate-preview-4.com) | 814.3 KiB | 15.5s | 3.1s |  | 2.7s | [Workflow run](https://github.com/iterate/iterate/actions/runs/275492288025158) | 2026-07-29T14:50:57.529Z | Preview app released. |
| Dummy Petshop | released | `d98251c` | [https://dummy-petshop.iterate-preview-4.com](https://dummy-petshop.iterate-preview-4.com) | 198.3 KiB | 53ms | 5.6s |  | 670ms | [Workflow run](https://github.com/iterate/iterate/actions/runs/275492288025158) | 2026-07-29T14:50:55.452Z | Preview app released. |
| OS | released | `d98251c` | [https://os.iterate-preview-4.com](https://os.iterate-preview-4.com) | 4.91 MiB (-10.3 KiB vs main) | 95.7s | 152.3s |  | 18.6s | [Workflow run](https://github.com/iterate/iterate/actions/runs/275492288025158) | 2026-07-29T14:51:13.346Z | Preview app released. |

</details>

### CODE REVIEW

File: apps/os/src/domains/email/utils.ts [hunk 1]
```diff
@@ -213,19 +213,20 @@ export function normalizeInboundEmailAllowedSender(pattern: string): string {
...
+ * `Authentication-Results: evil; dmarc=pass` alone no longer satisfies it.
  */
 export function dmarcPasses(authenticationResults: string | null): boolean {
   if (authenticationResults === null) return false;
-  return /\bdmarc=pass\b/i.test(authenticationResults);
+  // Pin authserv-id to Cloudflare's MX so a sender-forged AR header cannot
+  // satisfy the check by itself.
+  return /(?:^|[\r\n])[ \t]*mx\.cloudflare\.net\b[^\r\n]*\bdmarc=pass\b/i.test(
+    authenticationResults,
+  );
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/email/utils.ts]: Good catch — fixed in b448daf50.

`headers.get("authentication-results")` joins duplicates with `", "`, so newline-only matching could both:
1. miss a real `mx.cloudflare.net; dmarc=pass` after a forged line (comma join), and
2. false-match a forged `dmarc=pass` after a real CF fail because `[^\r\n]*` crossed the comma.

Now we split AR into records on newlines *and* commas that introduce a new authserv-id, then require `^mx.cloudflare.net` + `dmarc=pass` on the same record. Tests cover both newline- and comma-joined shapes.

File: apps/os/src/domains/email/utils.ts [hunk 2]
```diff
@@ -213,19 +213,28 @@ export function normalizeInboundEmailAllowedSender(pattern: string): string {
...
   if (authenticationResults === null) return false;
-  return /\bdmarc=pass\b/i.test(authenticationResults);
+  // Split into AR records. Fetch `Headers.get` joins multiples with ", ";
+  // raw MIME may separate them with newlines. Within a single record,
+  // methods are `;`-separated (not comma-separated).
+  const records = authenticationResults.split(/(?:\r?\n|,(?=\s*[^,;\s]+\s*;))/);
+  return records.some((record) => {
+    const trimmed = record.trim();
+    return /^mx\.cloudflare\.net\b/i.test(trimmed) && /\bdmarc=pass\b/i.test(trimmed);
+  });
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/email/utils.ts]: Fixed — `\b` after `net` really does accept `mx.cloudflare.net.evil` / `net-evil`.

Now the record must start with exact `mx.cloudflare.net` (optional version digit) then `;`, so subdomain/hyphen suffix forgeries no longer match. Added tests for those cases plus optional authserv-id version.

File: apps/os/src/domains/email/utils.ts [hunk 3]
```diff
@@ -213,19 +213,30 @@ export function normalizeInboundEmailAllowedSender(pattern: string): string {
...
+  // Split into AR records. Fetch `Headers.get` joins multiples with ", ";
+  // raw MIME may separate them with newlines. Within a single record,
+  // methods are `;`-separated (not comma-separated).
+  const records = authenticationResults.split(/(?:\r?\n|,(?=\s*[^,;\s]+\s*;))/);
+  return records.some((record) => {
+    const trimmed = record.trim();
+    // Exact authserv-id (optional version digit), then ";". A word-boundary
+    // pin would accept forgeries like mx.cloudflare.net.evil / net-evil.
+    return /^mx\.cloudflare\.net(?:\s+\d+)?\s*;/i.test(trimmed) && /\bdmarc=pass\b/i.test(trimmed);
+  });
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/email/utils.ts]: Fixed.

1. Comma is only an AR record boundary when the left side is already a complete `authserv-id; …` record — so `evil, mx.cloudflare.net; dmarc=pass` stays one record that does not start with CF and fails.
2. We take the **last** Cloudflare-authored record (MTA append order), so an earlier forged CF pass cannot beat a trailing real CF fail.

Tests cover the synthetic-comma forgery and last-CF-wins.

File: apps/os/src/domains/email/utils.ts [hunk 4]
```diff
@@ -213,19 +213,58 @@ export function normalizeInboundEmailAllowedSender(pattern: string): string {
...
+  let lastCloudflareRecord: string | undefined;
+  for (const record of splitAuthenticationResultsRecords(authenticationResults)) {
+    const trimmed = record.trim();
+    // Exact authserv-id (optional version digit), then ";". A word-boundary
+    // pin would accept forgeries like mx.cloudflare.net.evil / net-evil.
+    if (/^mx\.cloudflare\.net(?:\s+\d+)?\s*;/i.test(trimmed)) {
+      lastCloudflareRecord = trimmed;
+    }
+  }
+  return lastCloudflareRecord != null && /\bdmarc=pass\b/i.test(lastCloudflareRecord);
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/email/utils.ts]: Done — switched to `!!lastCloudflareRecord && …`.

File: apps/os/src/domains/email/utils.ts [hunk 5]
```diff
@@ -213,19 +213,58 @@ export function normalizeInboundEmailAllowedSender(pattern: string): string {
...
+    while ((match = re.exec(line)) !== null) {
+      const left = line.slice(start, match.index);
+      if (left.includes(";")) {
+        records.push(left);
+        start = match.index + 1;
+      }
+    }
+    records.push(line.slice(start));
+  }
+  return records;
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/email/utils.ts:270]: Fixed — the comma-boundary lookahead only accepted `authserv-id;` and missed optional version (`authserv-id 1;`), so a Headers.get join ending in a versioned CF record never split and legitimate mail could fail DMARC.

Lookahead is now `authserv-id (version)? ;`. Test added for `…, mx.cloudflare.net 1; dmarc=pass`.
